### PR TITLE
SSL geometry proto conversion

### LINKED ISSUE #1436: Simulator: Create geometry proto conversion functions
### Description of the task

In order for the new Simulator to output the same proto as SSL-Vision, we need functions to convert our internal datatypes like `Field` into the corresponding SSL geometry protos.

### Acceptance criteria

- [ ] as titled

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Added functions to convert our `Field` class into the corresponding `SSL_GeometryData` proto message. This will be later used for the Simulator to send out proto messages.

I added the conversion code in the same place as #1407 so they shouldn't conflict too much.

In addition to the conversion code I also updated the `Field` class. The new protos from #1433 actually added a new field with the `goal_depth`, so I needed to add that to the field class in order to be able to test everything properly.

I also updated some of the code in `SensorFusion` that converts the `SSL_GeometryData` into a `Field` to make use of my new "centralized" map to track line segment and arc names. Hopefully this makes things more clear.

### Testing Done

_lots and lots_ of unit tests (they hurt me)

### Resolved Issues

resolves #1436 

### Length Justification

It's 600 lines of tests alone, not much I can do to break it up. There was also a bit of refactoring required to update the `Field` class now that we have new proto data for the goal depth.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/proto/message_translation/ssl_message_translator_test.cpp
Comment (MathewMacDougall) [inline src/software/proto/message_translation/ssl_message_translator_test.cpp]: This will be resolved once #1418 is merged since it has some functions I need

File: src/software/proto/message_translation/ssl_message_translator.h
Comment (kelvinkoon) [inline src/software/proto/message_translation/ssl_message_translator.h:51]: Does the `MAKE_ENUM` file have methods to convert to strings so we can avoid using maps?

File: src/software/proto/message_translation/ssl_message_translator.cpp
Comment (kelvinkoon) [inline src/software/proto/message_translation/ssl_message_translator.cpp:94]: Is there any difference between using `throw std::invalid_argument` against using `assert`? Wanted to ask to keep consistency.

File: src/software/proto/message_translation/ssl_message_translator.h
Comment (kelvinkoon) [inline src/software/proto/message_translation/ssl_message_translator.h]: Does the `MAKE_ENUM` file have methods to convert to strings so we can avoid using maps? Or any other way we  can avoid this?

**Review by kelvinkoon — CHANGES_REQUESTED** [review]
Looks good overall, just checking a few things.

**Review by jonathanlew — CHANGES_REQUESTED** [review]
:pray: happy to see this coming along

Nice to have: an integration test i.e. 
```
Field expected = createSSLdivBfield
Field result = ssl_protobuf_reader.getFexpectedield(createGeometryData(createSSLdivBfield));
EXPECT_EQ(expected,result);
```

File: src/software/proto/message_translation/ssl_message_translator.cpp
Comment (MathewMacDougall) [reply src/software/proto/message_translation/ssl_message_translator.cpp:94]: We basically prefer exceptions in c++ code because they are recoverable (exceptions can be caught, assertions can not). Exceptions don't exist in C so we generally only use asserts there.

File: src/software/proto/message_translation/ssl_message_translator.h
Comment (MathewMacDougall) [reply src/software/proto/message_translation/ssl_message_translator.h]: Yes, but technically the convention for naming enums is `ALL_CAPS_WITH_UNDERSCORES`, however the strings _must_ be exact (and do not follow the same convention) which is why I did it this way.

I _could_ break enum convention and then use `MAKE_ENUM`. Thoughts @jonathanlew @garethellis0 @akhilveeraghanta @EvanMorcom

Comment (MathewMacDougall) [reply src/software/proto/message_translation/ssl_message_translator.h]: Hmmm I think that's only a `proto3` feature, I can't find a `descriptor` for the message types.

**Comment by MathewMacDougall** [comment]
> Nice to have: an integration test i.e.

Great idea. Added an integration test and fixed some bugs with converting units.

**Review by JNeiger — DISMISSED** [review]
Looks pretty solid!

File: src/software/proto/message_translation/ssl_message_translator.h
Comment (garethellis0) [reply src/software/proto/message_translation/ssl_message_translator.h]: I think I prefer following convention here, but also I don't believe these enum names should be explicitly tied to whatever names the SSL decides to use, we should be consistent internally.

File: src/software/proto/message_translation/ssl_message_translator.h
Comment (garethellis0) [inline src/software/proto/message_translation/ssl_message_translator.h]: Could we put the maps in the `.cpp` file and then just pass Enums into this? That we we totally encapsulate the fact that SSL uses raw strings in this way.

File: src/software/proto/message_translation/ssl_message_translator.h
Comment (garethellis0) [inline src/software/proto/message_translation/ssl_message_translator.h:119]: Similar comment here about using raw strings instead of an enum

Comment (garethellis0) [reply src/software/proto/message_translation/ssl_message_translator.h:119]: Similar comment applies for function(s) below.

File: src/software/world/field.h
Comment (garethellis0) [inline src/software/world/field.h]: including the buffer? just curious but I can (logically) see that being zero

**Review by garethellis0 — CHANGES_REQUESTED** [review]
nice job, looking forward to seeing this go in

**Review by jonathanlew — DISMISSED** [review]
Nice :bug: catch with the convert to field and back test!
Nothing more to add beyond what others have said

File: src/software/proto/message_translation/ssl_message_translator.h
Comment (MathewMacDougall) [reply src/software/proto/message_translation/ssl_message_translator.h]: Ok just to be clear, I'll rename the enums to try match our internal names / conventions as much as possible? We don't really have names for lines internally, just points (eg. `posXPosYCorner`). For example, I'm not super sure what "our version" of `TOP_TOUCH_LINE` would be. `POS_Y_FIELD_LINE`?

File: src/software/world/field.h
Comment (MathewMacDougall) [reply src/software/world/field.h]: Great point, I've changed it to allow boundary buffers of `0`

File: src/software/proto/message_translation/ssl_message_translator.h
Comment (MathewMacDougall) [reply src/software/proto/message_translation/ssl_message_translator.h]: done

File: src/software/proto/message_translation/ssl_message_translator.h
Comment (MathewMacDougall) [reply src/software/proto/message_translation/ssl_message_translator.h:119]: done for all functions that previously took a string

**Review by jonathanlew — APPROVED** [review]

**Review by kelvinkoon — APPROVED** [review]
:tada:

**Review by EvanMorcom — APPROVED** [review]
Very good, enum-ber one 🔥 🥇

File: src/software/proto/message_translation/ssl_message_translator.cpp
Comment (garethellis0) [reply src/software/proto/message_translation/ssl_message_translator.cpp:94]: It's worth noting as well that `asserts` should not generally be used for handling errors unless something _really_ badly goes wrong, because they will totally stop your program (which we generally try to avoid) and when we compile with all optimisations turned on, the compiler actually removes them all. They can be thought of as a helpful tool for debugging (much easier to see that an assert failed then see the effects of some invalid calculation downstream).

File: src/software/proto/message_translation/ssl_message_translator.h
Comment (garethellis0) [reply src/software/proto/message_translation/ssl_message_translator.h]: > I'm not super sure what "our version" of TOP_TOUCH_LINE would be. POS_Y_FIELD_LINE?

I like that

**Review by garethellis0 — APPROVED** [review]
LGTM pending existing comments :+1: :tada:

**Pull request merged**
